src/sinks/util/sink.rs

```diff
@@ -464,27 +469,33 @@ where
         let partition = item.partition();
 
         if let Some(batch) = self.partitions.get_mut(&partition) {
-            if batch.len() >= self.settings.size {
+            if batch.was_full() {
                 trace!("Batch full; driving service to completion.");
                 self.poll_complete()?;
 
                 if let Some(batch) = self.partitions.get_mut(&partition) {
-                    if batch.len() >= self.settings.size {
+                    if !batch.is_empty() {
                         debug!(
                             message = "Buffer full; applying back pressure.",
-                            max_size = %self.settings.size,
+                            max_bytes = %self.settings.bytes,
+                            max_events = %self.settings.events,
                             rate_limit_secs = 10
                         );
                         return Ok(AsyncSink::NotReady(item));
                     } else {
-                        batch.push(item);
-                        return Ok(AsyncSink::Ready);
+                        return match batch.push(item) {
+                            PushResult::Ok => Ok(AsyncSink::Ready),
+                            // FIXME is this the right behavior here and below?
+                            PushResult::Full(item) => Ok(AsyncSink::NotReady(item)),
```

Comment:
I don't think this is actually reachable since the batch should be empty. But either way it seems like the right behavior. There's no way to get here without calling `poll_complete` and then observing evidence that it did not return `Ready` (i.e. batch being no longer full).

Reply:
That's a relief.